### PR TITLE
fix(cpu-data-script): add attention mapping and context length config

### PR DESCRIPTION
**Pull request by prashantgupta24**
## Summary

  This PR enhances the `save_cpu_data.py` script with additional configuration options for attention mapping and context length management, improving compatibility with VLLM inference settings.

### Changes

  1. Attention Type Mapping (save_cpu_data.py:73-79)
    - Added a mapping dictionary to translate user-provided attention type arguments to internal attention names
    - FP8 detection now uses the mapped attention name for accurate identification
   
  2. Context Length Configuration (save_cpu_data.py:119-132)
    - Added dynamic calculation of required context length based on padded input length + max new tokens
    - Implements binary search (bisect) to find the smallest supported context length that accommodates the request
    - Supported context lengths: 64, 128, 256, 512, 1024, 2048, 4096, 8192, 16384, 32768
    - Sets `VLLM_DT_MAX_CONTEXT_LEN` environment variable accordingly
    
  3. Input Padding (save_cpu_data.py:108-115)
    - Added 64-block aligned padding for input IDs using `pad_input_ids` from `fms.utils.generation` (Similar to `inference.py`)
    - Ensures inputs are properly padded before being passed to the model
    
  4. VLLM Batch Size Tuning
    - Hard-coded `VLLM_DT_MAX_BATCH_SIZE` environment variable to 2 (4 was timing out)

### CODE REVIEW

File: aiu_fms_testing_utils/scripts/save_cpu_data.py [hunk 1]
```diff
@@ -67,9 +69,23 @@ def load_jsonl(path):
...
+}
+
 args = parser.parse_args()
 max_new_tokens = args.max_new_tokens
-is_fp8 = "fp8" in args.attention_type
+attn_name = attention_map[args.attention_type]
+is_fp8 = "fp8" in attn_name
+
+## Random defaults
+os.environ.setdefault("VLLM_DT_MAX_BATCH_SIZE", "4")
```
Comment (prashantgupta24) [inline aiu_fms_testing_utils/scripts/save_cpu_data.py]: seems to be working with 4, maybe needs to be set to something more accurate

File: aiu_fms_testing_utils/scripts/save_cpu_data.py [hunk 2]
```diff
@@ -67,9 +69,23 @@ def load_jsonl(path):
...
+}
+
 args = parser.parse_args()
 max_new_tokens = args.max_new_tokens
-is_fp8 = "fp8" in args.attention_type
+attn_name = attention_map[args.attention_type]
+is_fp8 = "fp8" in attn_name
+
+## Random defaults
+os.environ.setdefault("VLLM_DT_MAX_BATCH_SIZE", "2")
```
Comment (joerunde) [inline aiu_fms_testing_utils/scripts/save_cpu_data.py:86]: a bit confusing for me because I thought this is only picked up by the compiler to set the max batch size to compile to, but this script is only running on cpu to collect logits so nothing is compiled.

Does this also control for the batch size that gets run on cpu?

File: aiu_fms_testing_utils/scripts/save_cpu_data.py [hunk 3]
```diff
@@ -86,17 +102,44 @@ def load_jsonl(path):
...
+        512,
+        1024,
+        2048,
+        4096,
+        8192,
+        16384,
+        32768,
+    ]
+    idx = bisect.bisect_left(supported_context_lengths, largest_context)
+    os.environ["VLLM_DT_MAX_CONTEXT_LEN"] = str(supported_context_lengths[idx])
```
Comment (joerunde) [inline aiu_fms_testing_utils/scripts/save_cpu_data.py:137]: this will run out of bounds if largest_context > supported_context_lengths[-1], so a quick assertion beforehand would be nice

Comment (joerunde) [reply aiu_fms_testing_utils/scripts/save_cpu_data.py:137]: Other questions:

1. Should this use setdefault like VLLM_DT_MAX_BATCH_SIZE?
2. Is this env var actually picked up after `validation_model` has already been loaded?

File: aiu_fms_testing_utils/scripts/save_cpu_data.py [hunk 4]
```diff
@@ -67,9 +69,23 @@ def load_jsonl(path):
...
+    "paged_fp8": "spyre_paged_attn_fp8",
+}
+
 args = parser.parse_args()
 max_new_tokens = args.max_new_tokens
-is_fp8 = "fp8" in args.attention_type
+attn_name = attention_map[args.attention_type]
+is_fp8 = "fp8" in attn_name
+
+## Random defaults
```
Comment (joerunde) [inline aiu_fms_testing_utils/scripts/save_cpu_data.py]: could have an informative comment like "setting batch size to 2 to avoid CI timeouts on larger batches"

File: aiu_fms_testing_utils/scripts/save_cpu_data.py (hunk 2 above)
Comment (prashantgupta24) [reply aiu_fms_testing_utils/scripts/save_cpu_data.py:86]: Trying to remember but this was mostly needed by the `aiu_fms_testing_utils.utils.paged.generate` script, which seems to depend on a lot of variables that are set elsewhere

```
    # these variables are guaranteed to be set in another location (inference.py, test_decoders.py, etc.)
    # if we set these variables here, we run the risk of warming up and generating with different sizes
    _MAX_BATCH = int(os.environ["VLLM_DT_MAX_BATCH_SIZE"])
    _MAX_CONTEXT_LENGTH = int(os.environ["VLLM_DT_MAX_CONTEXT_LEN"])
```

File: aiu_fms_testing_utils/scripts/save_cpu_data.py (hunk 3 above)
Comment (prashantgupta24) [reply aiu_fms_testing_utils/scripts/save_cpu_data.py:137]: 1. No I don't think `setdefault` would work since we're setting it in a loop based on actual prompt length.
2. The `generate()` function in `paged.py` reads `os.environ["VLLM_DT_MAX_CONTEXT_LEN"]` which is what we're setting it for

Comment (joerunde) [reply aiu_fms_testing_utils/scripts/save_cpu_data.py:137]: ah gotcha, didn't see this was per-row. Very hot

**Review by joerunde — APPROVED** [review]
lpgtm!

File: aiu_fms_testing_utils/scripts/save_cpu_data.py [hunk 5]
```diff
@@ -86,17 +102,48 @@ def load_jsonl(path):
...
     prompt_text = row["prompt"]
     input_ids = tokenizer.encode(prompt_text)
     print("fetching cpu validation info for id: ", id)
+    BLOCK_SIZE = 64
+    prompt_len = len(input_ids)
+    padded_len = BLOCK_SIZE * ((prompt_len + BLOCK_SIZE - 1) // BLOCK_SIZE)
+    ids, extra_kwargs = pad_input_ids(
+        [torch.tensor(input_ids)], min_pad_length=padded_len
+    )
+    extra_kwargs["attn_name"] = attn_name
```
Comment (prashantgupta24) [inline aiu_fms_testing_utils/scripts/save_cpu_data.py:117]: A very important but subtle detail - by passing in `attn_name`, we change the function call in In the `extract_validation_information` function within `aiu_fms_testing_utils/testing/validation.py` script. We now effectively call `aiu_fms_testing_utils.utils.paged` which is the right implementation for `paged` attention.

```
    if "paged" in extra_kwargs.get("attn_name", "sdpa"):
        from aiu_fms_testing_utils.utils.paged import generate

        attention_specific_kwargs["prefill_chunk_size"] = prefill_chunk_size
    else:
        # TODO: Add a unified generation dependent on attn_type
        from fms.utils.generation import generate
```
